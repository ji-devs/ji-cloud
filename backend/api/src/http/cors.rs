use actix_web::http::{header, Method};
use core::config::CORS_ORIGINS;

pub fn get(local_insecure: bool) -> actix_cors::Cors {
    let mut cors = actix_cors::Cors::default()
        .supports_credentials()
        .allowed_methods(&[
            Method::GET,
            Method::POST,
            Method::PATCH,
            Method::DELETE,
            Method::PUT,
            Method::OPTIONS,
        ])
        .allowed_headers(&[
            header::AUTHORIZATION,
            header::CONTENT_TYPE,
            header::HeaderName::from_static("x-csrf"),
        ]);

    if local_insecure {
        cors = cors.allow_any_origin();
    } else {
        for origin in CORS_ORIGINS {
            cors = cors.allowed_origin(origin);
        }
    }

    cors
}
