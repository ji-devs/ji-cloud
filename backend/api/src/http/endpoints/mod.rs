pub mod additional_resource;
pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod module;
pub mod search;
pub mod session;
pub mod user;
