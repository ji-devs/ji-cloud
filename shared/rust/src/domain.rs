//! Types that travel over the wire.

macro_rules! into_uuid {
    ( $( $t:ty ),+ $(,)? ) => {
        $(
            impl From<$t> for uuid::Uuid {
                fn from(t: $t) -> Self {
                    t.0
                }
            }
        )+
    };
}

pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod search;
mod ser;
pub mod session;
pub mod user;

#[deprecated]
/// auth types (deprecated)
pub mod auth {

    #[deprecated]
    pub use super::session::AUTH_COOKIE_NAME;

    #[deprecated]
    pub use super::session::CSRF_HEADER_NAME;

    #[deprecated]
    pub use super::user::PutProfileRequest as RegisterRequest;
}

use chrono::Utc;
#[cfg(feature = "backend")]
use paperclip::actix::Apiv2Schema;
use ser::{csv_encode_uuids, deserialize_optional_field, from_csv};
use uuid::Uuid;

/// Serialize/Deserialize wrapper for Base64 encoded content.
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
#[cfg_attr(feature = "backend", openapi(empty))]
#[derive(Debug)]
pub struct Base64<T>(pub T);

impl<T: std::fmt::Display> serde::Serialize for Base64<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&base64::encode(&self.0.to_string()))
    }
}

impl<'de, E: std::fmt::Debug, T: std::str::FromStr<Err = E>> serde::Deserialize<'de> for Base64<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Ok(Self(deserializer.deserialize_str(ser::FromStrVisiter(
            std::marker::PhantomData,
        ))?))
    }
}
/// Response for successfuly creating a Resource.
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub struct CreateResponse<T: Into<Uuid>> {
    /// The newly created resource's ID.
    pub id: T,
}

/// Represents when to publish an image.
#[derive(Copy, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(Apiv2Schema))]
pub enum Publish {
    /// Publish the image *at* the given time.
    At(chrono::DateTime<Utc>),
    /// Publish the image *in* the given amount of time from now.
    In(std::time::Duration),
}

impl Publish {
    /// creates an instance of `Self` that will publish "right now"
    #[must_use]
    #[allow(clippy::missing_const_for_fn)]
    pub fn now() -> Self {
        // Duration::new is const unstable
        Self::In(std::time::Duration::new(0, 0))
    }
}

impl From<Publish> for chrono::DateTime<Utc> {
    fn from(publish: Publish) -> Self {
        match publish {
            Publish::At(t) => t,
            Publish::In(d) => {
                // todo: error instead of panicing
                Utc::now() + chrono::Duration::from_std(d).expect("Really really big duration?")
            }
        }
    }
}
