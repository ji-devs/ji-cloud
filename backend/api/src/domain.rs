use actix_http::body::BoxBody;
use actix_web::{
    cookie::{Cookie, SameSite},
    HttpResponse, Responder,
};

use crate::{
    db::{self, account::AccountMember},
    error,
};
use chrono::{DateTime, Utc};
use http::StatusCode;
use shared::domain::billing::AccountId;
use shared::domain::user::{UserId, UserScope};
use shared::domain::{
    category::{Category, CategoryId},
    session::AUTH_COOKIE_NAME,
};
use sqlx::PgPool;
use std::{cell::RefCell, convert::TryFrom, rc::Rc};
use std::{collections::HashMap, fmt};
use tracing::instrument;
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
}

#[derive(Debug, Clone)]
struct CategoryNode {
    pub id: Uuid,
    pub name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
    pub children: Vec<Rc<RefCell<CategoryNode>>>,
}

impl From<CategoryNode> for Category {
    fn from(category_node: CategoryNode) -> Self {
        Self {
            id: CategoryId(category_node.id),
            name: category_node.name,
            created_at: category_node.created_at,
            updated_at: category_node.updated_at,
            user_scopes: {
                category_node
                    .user_scopes
                    .into_iter()
                    .map(|x| UserScope::try_from(x).expect("detected an invalid user scope"))
                    .collect()
            },
            children: {
                category_node
                    .children
                    .iter()
                    .map(|category| {
                        let category: CategoryNode = (**category).borrow().clone();
                        category.into()
                    })
                    .collect()
            },
        }
    }
}

#[instrument(skip_all)]
pub fn build_tree(categories: Vec<RawCategory>) -> Vec<Category> {
    let mut nodes: Vec<Rc<RefCell<CategoryNode>>> = Vec::new();
    let mut lookup: HashMap<Uuid, Rc<RefCell<CategoryNode>>> = HashMap::new();

    // Now we know this category exists in the lookup table
    for raw in categories.iter() {
        lookup.insert(
            raw.id.clone(),
            Rc::new(RefCell::new(CategoryNode {
                id: raw.id.clone(),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                user_scopes: raw.user_scopes.clone(),
                children: Vec::new(),
            })),
        );
    }

    for raw in categories.iter() {
        let current_node = lookup
            .get(&raw.id)
            .expect("Category wasn't added to lookup map");
        match raw.parent_id {
            None => {
                // This is a root category, add it to the nodes list
                nodes.push(current_node.clone());
            }
            Some(parent_id) => {
                let parent_node = lookup
                    .get(&parent_id)
                    .expect("Parent caregory wasn't added to lookup map");
                let mut parent_node = (**parent_node).borrow_mut();
                parent_node.children.push(current_node.clone());
            }
        }
    }

    nodes
        .iter()
        .map(|node| {
            let category: CategoryNode = (**node).borrow().clone();
            let category: Category = category.into();
            category
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}

#[derive(Debug)]
pub struct NoContentClearAuth;

impl fmt::Display for NoContentClearAuth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("No Content")
    }
}

impl Responder for NoContentClearAuth {
    type Body = BoxBody;

    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse<BoxBody> {
        let mut cookie = Cookie::named(AUTH_COOKIE_NAME);
        cookie.set_max_age(time::Duration::seconds(0));
        cookie.set_http_only(true);
        cookie.set_same_site(SameSite::Strict);

        actix_web::HttpResponse::build(StatusCode::NO_CONTENT)
            .content_type("application/octet-stream")
            .cookie(cookie)
            .finish()
    }
}

#[derive(Debug)]
pub enum UserAuthorization {
    SystemAdministrator,
    AccountAdministrator,
    AccountMember,
}

impl UserAuthorization {
    pub fn is_system_administrator(&self) -> bool {
        matches!(self, UserAuthorization::SystemAdministrator)
    }

    pub fn test_authorized(&self, require_account_admin: bool) -> Result<(), error::Account> {
        if self.is_authorized(require_account_admin) {
            Ok(())
        } else {
            Err(error::Account::Forbidden)
        }
    }

    pub fn is_authorized(&self, require_account_admin: bool) -> bool {
        match self {
            UserAuthorization::AccountMember if require_account_admin => false,
            _ => true,
        }
    }
}

pub async fn user_authorization(
    db: &PgPool,
    user_id: &UserId,
    account_id: &AccountId,
) -> Result<UserAuthorization, error::Account> {
    Ok(
        if db::user::has_scopes(db, *user_id, &[UserScope::Admin]).await? {
            UserAuthorization::SystemAdministrator
        } else {
            match db::account::user_account_membership(db, user_id, account_id).await? {
                Some(AccountMember::Admin) => UserAuthorization::AccountAdministrator,
                Some(AccountMember::User) => UserAuthorization::AccountMember,
                None => return Err(error::Account::Forbidden),
            }
        },
    )
}
