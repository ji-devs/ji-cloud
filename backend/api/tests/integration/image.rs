mod recent;
mod tag;
mod user;

use http::StatusCode;
use serde_json::json;
use shared::domain::{image::ImageId, CreateResponse};
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
use uuid::Uuid;

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
    service::Service,
};

async fn create(
    styles: &[Uuid],
    age_ranges: &[Uuid],
    affiliations: &[Uuid],
    categories: &[Uuid],
    tags: &[i16],
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(
        &[Fixture::User, Fixture::Image, Fixture::MetaKinds],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": (),
            "styles": styles,
            "age_ranges": age_ranges,
            "affiliations": affiliations,
            "categories": categories,
            "tags": tags,
            "size": "Canvas",
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<ImageId> = resp.json().await?;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    Ok(())
}

#[sqlx::test]
async fn create_no_meta(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create(&[], &[], &[], &[], &[], pool_opts, conn_opts).await
}

#[sqlx::test]
async fn create_with_styles(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create(
        &[
            "6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?,
            "6389ff7c-de76-11ea-b7ab-9b5661dd4f70".parse()?,
        ],
        &[],
        &[],
        &[],
        &[],
        pool_opts,
        conn_opts,
    )
    .await
}

#[sqlx::test]
async fn create_with_meta(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create(
        &["6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?],
        &["f3722790-de76-11ea-b7ab-77b45e9af3ef".parse()?],
        &["c0cd4446-de76-11ea-b7ab-93987e8aa112".parse()?],
        &[],
        &[1],
        pool_opts,
        conn_opts,
    )
    .await
}

async fn create_error(
    kind: &str,
    id: &str,
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[], pool_opts, conn_opts).await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": null,
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "categories": [],
            "tags": [],
            "size": "Canvas",
            kind: [id],
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNPROCESSABLE_ENTITY);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

async fn create_error_tag(
    kind: &str,
    id: &i16,
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[], pool_opts, conn_opts).await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": null,
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "categories": [],
            "tags": [],
            "size": "Canvas",
            kind: [id],
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNPROCESSABLE_ENTITY);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[sqlx::test]
async fn create_with_style_error(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create_error(
        "styles",
        "6389eaa0-de76-11ea-b7ab-0399bcf84df2",
        pool_opts,
        conn_opts,
    )
    .await
}

#[sqlx::test]
async fn create_with_affiliation_error(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create_error(
        "affiliations",
        "6389eaa0-de76-11ea-b7ab-0399bcf84df2",
        pool_opts,
        conn_opts,
    )
    .await
}

#[sqlx::test]
async fn create_with_age_range_error(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create_error(
        "age_ranges",
        "6389eaa0-de76-11ea-b7ab-0399bcf84df2",
        pool_opts,
        conn_opts,
    )
    .await
}

#[sqlx::test]
async fn create_with_category_error(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create_error(
        "categories",
        "6389eaa0-de76-11ea-b7ab-0399bcf84df2",
        pool_opts,
        conn_opts,
    )
    .await
}

#[sqlx::test]
async fn create_with_tags_error(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    create_error_tag("tags", &22, pool_opts, conn_opts).await
}

#[sqlx::test]
async fn get_metadata(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    let app = initialize_server(
        &[
            Fixture::User,
            Fixture::MetaKinds,
            Fixture::Image,
            Fixture::MetaImage,
        ],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

// todo: search; missing algolia, s3
// todo: upload; missing algolia, s3
// todo: delete; missing algolia, s3
// todo: delete: edge case (never uploaded, should work even without s3), missing algolia

async fn update(
    req: &serde_json::Value,
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(
        &[Fixture::User, Fixture::MetaKinds, Fixture::Image],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .json(req)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

#[sqlx::test]
async fn update_empty(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    update(&json!({}), pool_opts, conn_opts).await
}

#[sqlx::test]
async fn update_is_premium(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    update(&json!({"is_premium": true}), pool_opts, conn_opts).await
}

#[sqlx::test]
async fn update_styles(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    update(&json!({"styles": ["6389eaa0-de76-11ea-b7ab-0399bcf84df2", "6389ff7c-de76-11ea-b7ab-9b5661dd4f70"]}), pool_opts,
conn_opts
    )
    .await
}

#[sqlx::test]
async fn update_tags(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    update(&json!({"tags": [0, 2]}), pool_opts, conn_opts).await
}

#[sqlx::test]
async fn browse(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image], &[], pool_opts, conn_opts).await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    // create a new image resource
    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/image/browse", port))
        .query(&[("page", "0"), ("size", "Canvas")])
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

// https://cloud.google.com/storage/docs/performing-resumable-uploads#single-chunk-upload
#[ignore]
#[sqlx::test]
async fn upload_with_url(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let file: Vec<u8> = include_bytes!("../../fixtures/images/ji-logo.png").to_vec();

    let app = initialize_server(
        &[Fixture::User, Fixture::Image],
        &[Service::GoogleCloudStorage],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    // get an upload URL for the new media
    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/8cca720a-c4bb-11eb-8edf-63da1d86939c/raw",
            port,
        ))
        .json(&json!({ "file_size": &file.len() }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let resp: shared::domain::image::ImageUploadResponse = resp.json().await?;

    let resp = client
        .put(&resp.session_uri)
        .header(reqwest::header::CONTENT_LENGTH, &file.len().to_string())
        .body(file)
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::OK);

    Ok(())
}

#[ignore]
#[sqlx::test]
async fn create_media_and_upload_with_url(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image], &[], pool_opts, conn_opts).await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    // create a new image resource
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .json(&json!({
            "name": "test_img",
            "description": "test_descip",
            "is_premium": false,
            "publish_at": (),
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "tags": [],
            "categories": [],
            "size": "Canvas",
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(StatusCode::CREATED, resp.status());

    let resp: shared::domain::image::CreateResponse = resp.json().await?;

    let id = resp.id.0;

    // get an upload URL for the new media
    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/{}/raw",
            port,
            id.to_string()
        ))
        .json(&json!({"file_size": 5}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let resp: shared::domain::image::ImageUploadResponse = resp.json().await?;

    let _url = resp.session_uri;

    // perform upload in single chunk
    let _resp = client;

    Ok(())
}
