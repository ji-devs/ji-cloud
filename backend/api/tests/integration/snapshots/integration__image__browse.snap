---
source: tests/integration/image.rs
expression: body

---
{
  "images": [
    {
      "metadata": {
        "id": "92d8c2d0-0aa4-11ec-9f5e-a3eea787c2a0",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "8cca6f3a-c4bb-11eb-8edf-13c75672da8f",
        "name": "test1",
        "description": "test description 1",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "8cca7124-c4bb-11eb-8edf-7b42383ed8f5",
        "name": "test2",
        "description": "test description 2",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "8cca719c-c4bb-11eb-8edf-f7accb638a15",
        "name": "test3",
        "description": "test description ",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710043d0-0a9b-11ec-9f5e-47290fe78365",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "7100447a-0a9b-11ec-9f5e-27e93628ecf8",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710044ca-0a9b-11ec-9f5e-af0642589fb8",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "7100451a-0a9b-11ec-9f5e-17d67fa55c21",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004560-0a9b-11ec-9f5e-130e0f157cd2",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710045a6-0a9b-11ec-9f5e-ab6585077bcd",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710045ec-0a9b-11ec-9f5e-4705e4763c4a",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004632-0a9b-11ec-9f5e-53c3e963010b",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004678-0a9b-11ec-9f5e-0f321f64bed2",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710046b4-0a9b-11ec-9f5e-cb07bd3af092",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710046fa-0a9b-11ec-9f5e-9b77f8801041",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004740-0a9b-11ec-9f5e-ef9aad54a89e",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004786-0a9b-11ec-9f5e-c3aad9c7738a",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "710047c2-0a9b-11ec-9f5e-93af7ac3e44c",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "71004808-0a9b-11ec-9f5e-27b95bb4a925",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    },
    {
      "metadata": {
        "id": "7100484e-0a9b-11ec-9f5e-e7d60617d001",
        "name": "test",
        "description": "test description",
        "is_premium": false,
        "kind": "Canvas",
        "publish_at": null,
        "styles": [],
        "tags": [],
        "age_ranges": [],
        "affiliations": [],
        "categories": [],
        "created_at": "2021-05-01T18:06:31.575087Z",
        "updated_at": null
      }
    }
  ],
  "pages": 2,
  "totalImageCount": 36
}
